core/src/main/java/com/sequenceiq/cloudbreak/service/decorator/TemplateDecorator.java

```diff
@@ -47,28 +47,33 @@
     public Template decorate(Credential credential, Template template, String region, String availabilityZone, String variant, CdpResourceType cdpResourceType) {
         setRootVolumeSize(template);
         PlatformDisks platformDisks = cloudParameterService.getDiskTypes();
-        CloudVmTypes vmTypesV2 = cloudParameterService.getVmTypesV2(
-                extendedCloudCredentialConverter.convert(credential),
-                region,
-                variant,
-                cdpResourceType,
-                new HashMap<>());
-        String locationString = locationService.location(region, availabilityZone);
-        VolumeParameterConfig config;
-        for (VolumeTemplate volumeTemplate : template.getVolumeTemplates()) {
-            try {
-                config = resolveVolumeParameterConfig(template, volumeTemplate, platformDisks, vmTypesV2, locationString);
-            } catch (NoSuchElementException ignored) {
-                LOGGER.debug("No VolumeParameterConfig found, which might be normal for platforms like YARN");
-                config = VolumeParameterConfig.EMPTY;
-            }
-
-            if (config.volumeParameterType() != null) {
-                if (volumeTemplate.getVolumeCount() == null) {
-                    volumeTemplate.setVolumeCount(config.maximumNumber());
+        boolean needToFetchTemplateCountAndSize = template.getVolumeTemplates().stream()
+                .anyMatch(it -> it.getVolumeCount() == null || it.getVolumeSize() == null);
+        if (needToFetchTemplateCountAndSize) {
+            LOGGER.info("We need to fetch the VM types. Perhaps the client does not provide the volume count or size. {}", template.getVolumeTemplates());
+            CloudVmTypes vmTypesV2 = cloudParameterService.getVmTypesV2(
+                    extendedCloudCredentialConverter.convert(credential),
+                    region,
+                    variant,
+                    cdpResourceType,
+                    new HashMap<>());
+            String locationString = locationService.location(region, availabilityZone);
+            VolumeParameterConfig config;
+            for (VolumeTemplate volumeTemplate : template.getVolumeTemplates()) {
```

Comment:
Not a big deal but I would rather move the variable def inside the loop body as it is not needed after the loop execution. (And it also gets overwritten in each loop iteration anyway.)